base/reflection.jl

```diff
@@ -287,28 +287,36 @@ isbits(t::DataType) = (@_pure_meta; !t.mutable & (t.layout != C_NULL) && datatyp
 isbits(t::Type) = (@_pure_meta; false)
 isbits(x) = (@_pure_meta; isbits(typeof(x)))
 
+_isleaftype(@nospecialize(t)) = (@_pure_meta; isa(t, DataType) && t.isleaftype)
+
 """
-    isleaftype(T)
+    isconcrete(T)
 
-Determine whether `T`'s only subtypes are itself and `Union{}`. This means `T` is
-a concrete type that can have instances.
+Determine whether `T` is a concrete type, meaning it can have direct instances
+(values `x` such that `typeof(x) === T`).
 
 # Examples
 ```jldoctest
-julia> isleaftype(Complex)
+julia> isconcrete(Complex)
 false
 
-julia> isleaftype(Complex{Float32})
+julia> isconcrete(Complex{Float32})
 true
 
-julia> isleaftype(Vector{Complex})
+julia> isconcrete(Vector{Complex})
 true
 
-julia> isleaftype(Vector{Complex{Float32}})
+julia> isconcrete(Vector{Complex{Float32}})
 true
+
+julia> isconcrete(Union{})
+false
+
+julia> isconcrete(Union{Int,String})
+false
 ```
 """
-isleaftype(@nospecialize(t)) = (@_pure_meta; isa(t, DataType) && t.isleaftype)
+isconcrete(@nospecialize(t)) = (@_pure_meta; isa(t, DataType) && t.layout != C_NULL)
```

Comment:
`layout != C_NULL` merely means that we were able to precompute something about the possible layouts of it's concrete subtypes. It's an even more complicated and ill-defined heuristic than `isleaftype` was.

Reply:
Maybe `t.layout != C_NULL && t.isleaftype` covers it?

Reply:
I attempted to pull together a list of the options, and posted in in the [original issue](https://github.com/JuliaLang/julia/issues/17086#issuecomment-328912370) to try to avoid forking the original conversation.